File: ueueo-framework/ueueo-background-jobs/src/main/java/com/ueueo/backgroundjobs/InMemoryBackgroundJobStore.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Volo.Abp.DependencyInjection;
using Volo.Abp.Timing;

namespace Volo.Abp.BackgroundJobs;

public class InMemoryBackgroundJobStore : IBackgroundJobStore, ISingletonDependency
{
    private readonly ConcurrentDictionary<Guid, BackgroundJobInfo> _jobs;

    protected IClock Clock;//  { get; }

    /**
     * Initializes a new instance of the <see cref="InMemoryBackgroundJobStore"/> class.
    */
    public InMemoryBackgroundJobStore(IClock clock)
    {
        Clock = clock;
        _jobs = new ConcurrentDictionary<Guid, BackgroundJobInfo>();
    }

    public   Task<BackgroundJobInfo> FindAsync(Guid jobId)
    {
        return Task.FromResult(_jobs.GetOrDefault(jobId));
    }

    public void InsertAsync(BackgroundJobInfo jobInfo)
    {
        _jobs[jobInfo.Id] = jobInfo;

        return Task.FromResult(0);
    }

    public   Task<List<BackgroundJobInfo>> GetWaitingJobsAsync(int maxResultCount)
    {
        var waitingJobs = _jobs.Values
            .Where(t => !t.IsAbandoned && t.NextTryTime <= Clock.Now)
            .OrderByDescending(t => t.Priority)
            .ThenBy(t => t.TryCount)
            .ThenBy(t => t.NextTryTime)
            .Take(maxResultCount)
            .ToList();

        return Task.FromResult(waitingJobs);
    }


    public void DeleteAsync(Guid jobId)
    {
        _jobs.TryRemove(jobId, out _);

        return Task.FromResult(0);
    }

    public void UpdateAsync(BackgroundJobInfo jobInfo)
    {
        if (jobInfo.IsAbandoned)
        {
            return DeleteAsync(jobInfo.Id);
        }

        return Task.FromResult(0);
    }
}

File: ueueo-framework/ueueo-background-jobs/src/main/java/com/ueueo/backgroundjobs/InMemoryBackgroundJobStore.java
package com.ueueo.backgroundjobs;

import com.ueueo.ID;

import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class InMemoryBackgroundJobStore implements IBackgroundJobStore {
    private ConcurrentHashMap<ID, BackgroundJobInfo> jobs;

    /**
     * Initializes a new instance of the <see cref="InMemoryBackgroundJobStore"/> class.
     */
    public InMemoryBackgroundJobStore() {
        jobs = new ConcurrentHashMap<>();
    }

    @Override
    public BackgroundJobInfo find(ID jobId) {
        return jobs.get(jobId);
    }

    @Override
    public void insert(BackgroundJobInfo jobInfo) {
        jobs.put(jobInfo.getId(), jobInfo);
    }

    @Override
    public List<BackgroundJobInfo> getWaitingJobs(int maxResultCount) {
        List<BackgroundJobInfo> waitingJobs = jobs.values().stream()
                .filter(t -> !t.isAbandoned() && t.getNextTryTime().before(new Date()))
                .sorted(new Comparator<BackgroundJobInfo>() {
                    @Override
                    public int compare(BackgroundJobInfo job1, BackgroundJobInfo job2) {
                        if (job1.getPriority().getIntValue() > job2.getPriority().getIntValue()) {
                            return -1;
                        } else if (job1.getPriority().getIntValue() < job2.getPriority().getIntValue()) {
                            return 1;
                        }
                        if (job1.getTryCount() > job2.getTryCount()) {
                            return -1;
                        } else if (job1.getTryCount() < job2.getTryCount()) {
                            return 1;
                        }
                        if (job1.getNextTryTime().after(job2.getNextTryTime())) {
                            return -1;
                        } else if (job1.getNextTryTime().before(job2.getNextTryTime())) {
                            return 1;
                        }
                        return 0;
                    }
                })
                .limit(maxResultCount)
                .collect(Collectors.toList());

        //        List<BackgroundJobInfo> waitingJobs = _jobs.Values
        //            .Where(t => !t.IsAbandoned && t.NextTryTime <= Clock.Now)
        //            .OrderByDescending(t => t.Priority)
        //            .ThenBy(t => t.TryCount)
        //            .ThenBy(t => t.NextTryTime)
        //            .Take(maxResultCount)
        //            .ToList();

        return waitingJobs;
    }

    @Override
    public void delete(ID jobId) {
        jobs.remove(jobId);

    }

    @Override
    public void update(BackgroundJobInfo jobInfo) {
        if (jobInfo.isAbandoned()) {
            delete(jobInfo.getId());
        }

    }
}
